Add write support for kafka connector

charlesjmorgan opened the pull request:
Add write support for the kafka connector
Add encoder to serialize message into avro, csv, json, and raw formats (works for primitives and json date/time types)
Currently some changes proposed in #4183 are included in this pr, but once those get merged I'll rebase (done)
Closes #3980

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,129 @@
+                messageEncoder);
+    }
+
+    private Map<String, String> getEncoderParameters(Optional<String> dataSchema)
```
dataSchema -> dataSchemaLocation

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/EncoderColumnHandle.java:
```diff
@@ -0,0 +1,33 @@
+public interface EncoderColumnHandle
+        extends ColumnHandle
+{
+    boolean isInternal();
```
We can drop this.

charlesjmorgan replied:
This method is used in `KafkaPageSinkProvider#createPageSink` to check for internal columns

aalbu replied:
Of course, I don't know why I suggested that.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroRowEncoder.java:
```diff
@@ -0,0 +1,71 @@
+        rowData.forEach(fieldValueProvider -> columnEncoders.get(fieldValueProvider.getColumn()).encodeField(fieldValueProvider, record));
+        try (DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(avroRecordWriter)) {
+            dataFileWriter.create(schema, outputStream);
+            dataFileWriter.append(record);
```
Perhaps we should consider changing the `RowEncoder` interface to return a `ProducerRecord` instead of a `byte[]`.  That would allow us to just create a `GenericData.Record` and let the Kafka producer worry about serialization.  Seems like a safer approach.  WDYT?

charlesjmorgan replied:
Would that make the encoder kafka specific?

aalbu replied:
It lives in the Kafka module, right :)  I know you envisioned this as a parallel to `RecordDecoder`, but it's not at this point and it shouldn't prevent us from making the right decision here.  I don't know if `RowEncoder` can be a connector-agnostic interface - I don't know for how many connectors a `byte[]` is sufficient for writing.  And we really wouldn't gain much by having this be connector-agnostic.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvColumnEncoder.java:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    private boolean isSupportedType(Type type)
```
This method is unnecessary.  It always produces the same result as `isSupportedPrimitive()`.  Rename the other method to `isSupportedType` and drop this one.

charlesjmorgan replied:
I had that there in case we add complex types, but I guess that is unlikely for csv format

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,68 @@
+            }
+            writer.writeNext(row);
+            writer.flush();
+            return csvStr.toString().getBytes(UTF_8);
```
It's more efficient to use a `ByteArrayOutputStream` as a sink and avoid creating a string.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,90 @@
+        try (KafkaProducer<byte[], byte[]> producer = producerFactory.create()) {
+            List<EncoderFieldValueProvider> row = new ArrayList<>(columnTypes.size());
+            for (int position = 0; position < page.getPositionCount(); ++position) {
+                for (int channel = 0; channel < page.getChannelCount(); ++channel) {
```
I must still be in c++ mode from school haha

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,129 @@
+                handle.getKeyDataFormat(),
+                getEncoderParameters(handle.getKeyDataSchemaLocation()),
+                handle.getColumns().stream()
+                        .filter(col -> !col.isInternal())
```
Actually, we probably have to be selective in the treatment of internal columns.  When publishing, one can specify the partition the message should go to.  So if the column list includes `_partition_id`, perhaps we should pass that to the Kafka producer.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,88 @@
+            List<EncoderFieldValueProvider> row = new ArrayList<>(columnTypes.size());
+            for (int position = 0; position < page.getPositionCount(); ++position) {
+                for (int channel = 0; channel < page.getChannelCount(); ++channel) {
+                    row.add(new KafkaFieldValueProvider(columns.get(channel), page.getBlock(channel), position));
```
We should segregate key columns (if any) and use the `keyEncoder` to generate the key.  You can probably tackle this and dealing with internal columns as a follow-up PR (no need to make this one larger than it already is).

charlesjmorgan commented:
Made revisions suggested by @aalbu 
Changed `RowEncoder#encodeRow` to return a `ProducerRecord`
Added serializers for different data formats
Removed encoder tests that don't work anymore (will add more in the future)
Improved round trip test design in `TestKafkaIntegrationSmokeTest`
Rebased and brought in bug fix changes for the `RawRowDecoder`

findepi commented:
cc @elonazoulay

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInsertTableHandle.java:
```diff
@@ -0,0 +1,153 @@
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.keyDataFormat = requireNonNull(keyDataFormat, "keyDataFormat is null");
+        this.messageDataFormat = requireNonNull(messageDataFormat, "messageDataFormat is null");
+        this.keyDataSchemaLocation = keyDataSchemaLocation;
```
requireNonNull

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInsertTableHandle.java:
```diff
@@ -0,0 +1,153 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(schemaName, tableName, topicName);
```
Why only three fields are used here? Also do you need hashCode/equals at all?

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaOutputTableHandle.java:
```diff
@@ -0,0 +1,153 @@
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaOutputTableHandle
```
What is the difference between `KafkaInsertTableHandle`? Maybe one class could be used for both?

charlesjmorgan replied:
Agreed, made both early on because I thought I'd need them

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/CustomDateTimeJsonFieldEncoder.java:
```diff
@@ -0,0 +1,98 @@
+            node.put(columnName, formatter.print(Instant.ofEpochMilli(DAYS.toMillis(fieldValueProvider.getLong())).toDateTime()));
+        }
+        else if (columnType == TIME) {
+            node.put(columnName, formatter.print(Instant.ofEpochMilli(fieldValueProvider.getLong()).toDateTime()));
```
I noticed that timestamp and timestamp_with_timezone were deprecated so I wasn't entirely sure what to do with them

findepi replied:
the constants are deprecated, but the types are not.
the deprecation is result of making them parametric, and so a singleton is passe

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/dummy/DummyRowEncoder.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    public ProducerRecord<Object, Object> encodeRow(List<EncoderFieldValueProvider> rowData, String topicName)
+    {
+        return null;
```
If the user doesn't specify an encoder to use in the table description it uses this one. This could use some thought

losipiuk replied:
I would the insert to fail in such case.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
```
should the producer be a field of `KafkaPageSink`? It gets created in a try with resources in `KafkaPageSink#appendPage`.

aalbu replied:
Right, we should associate the producer with the sink, not each page.

findepi commented on presto-kafka/pom.xml:
```diff
@@ -191,5 +206,12 @@
         </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
```
https://github.com/prestosql/presto/pull/648#pullrequestreview-234652050

> Regarding usage of Mockito, we have a general policy of not using mocking libraries in Presto, as they tend to encourage bad practices when writing code, and they often result in brittle tests that depend on internal implementation details rather than external behavior.

kokosing replied:
https://github.com/prestosql/presto/wiki/Code-style#mocking

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/EncoderColumnHandle.java:
```diff
@@ -0,0 +1,33 @@
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.type.Type;
+
+public interface EncoderColumnHandle
```
This interface is same as `DecoderColumnHandle`.
While I see some reasons for this on my own, i would like to learn your reasoning behind the decision to have them as 2 separate.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroRowEncoder.java:
```diff
@@ -0,0 +1,67 @@
+        implements RowEncoder
+{
+    public static final String NAME = "avro";
+    private final DatumWriter<GenericRecord> avroRecordWriter;
```
unused

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/PlainTextKafkaProducerFactory.java:
```diff
@@ -0,0 +1,58 @@
+    public PlainTextKafkaProducerFactory(KafkaConfig kafkaConfig)
+    {
+        requireNonNull(kafkaConfig, "kafkaConfig is null");
+        nodes = kafkaConfig.getNodes();
```
defensive copy

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/PlainTextKafkaProducerFactory.java:
```diff
@@ -0,0 +1,58 @@
+import static org.apache.kafka.clients.producer.ProducerConfig.LINGER_MS_CONFIG;
+import static org.apache.kafka.clients.producer.ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG;
+
+public class PlainTextKafkaProducerFactory
```
what's plain text about this class?
please add a comment what is this named after

i guess this is modelled this way after `PlainTextKafkaConsumerFactory`, so -- cc @kokosing

charlesjmorgan replied:
You're right I modeled it after the `PlainTextKafkaConsumerFactory` although I'm not sure what that name means. I was trying to think of a different name for it on friday but couldn't come up with one

kokosing replied:
PlainText is about security. This is how it is named in kafka. https://stackoverflow.com/a/50742123/217852

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaProducerFactory.java:
```diff
@@ -0,0 +1,28 @@
+        return new KafkaProducer<>(configure(keySerializer, messageSerializer));
+    }
+
+    Properties configure(String keySerializer, String messageSerializer);
```
Should it be part of interface?

Since we have 1 implementation, do we need the interface at all?

charlesjmorgan replied:
I don't think we need the interface unless we want to create producers with different properties. All I can think of for this is maybe different producers for different size inserts? Like if it's a very large insert we could change the batch size or buffer?

findepi replied:
There is no way to pass `Properties` _in_ (and no way to know what valid prop keys there are).
You seem to have the method as a helper to the default method. A regular class, and a protected methodn would be more suitable.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -0,0 +1,25 @@
+{
+    ProducerRecord<Object, Object> encodeRow(List<EncoderFieldValueProvider> rowData, String topicName);
+
+    String getSerializer();
```
Instead of passing class by name, we could pass actual objects, if we invoked `org.apache.kafka.clients.producer.KafkaProducer#KafkaProducer(java.util.Properties, org.apache.kafka.common.serialization.Serializer<K>, org.apache.kafka.common.serialization.Serializer<V>)` -- `KafkaProducer` constructor overload.

The gain is not in terms of saving reflective call for performance reasons, but to increase flow visibility within IDE.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInsertTableHandle.java:
```diff
@@ -0,0 +1,153 @@
+        return columns;
+    }
+
+    public SchemaTableName toSchemaTableName()
```
unused?

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFieldValueProvider.java:
```diff
@@ -0,0 +1,172 @@
+        return Objects.hash(columnHandle, block, position);
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        KafkaFieldValueProvider other = (KafkaFieldValueProvider) obj;
+        return Objects.equals(this.columnHandle, other.columnHandle)
+                && Objects.equals(this.block, other.block)
+                && Objects.equals(this.position, other.position);
+    }
```
does this class need equals?

block,position -based equality is not really value-based equality, so i suspect it does not

charlesjmorgan replied:
no it doesn't, I never use it

findepi replied:
So please remove it.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFieldValueProvider.java:
```diff
@@ -0,0 +1,172 @@
+ *
+ */
+public class KafkaFieldValueProvider
+        implements EncoderFieldValueProvider
```
we don't need the interface, since encoders are currently kafka-only
also, since the code here is not really kafka-related, you can call the class `EncoderFieldValueProvider`

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFieldValueProvider.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    @Override
+    public float getReal()
```
Rethink naming in this class. It's called 'real' after `REA` SQL type (otherwise you would call it `getFloat`).
But `getShort` is called after `short` -- java representation for the `smallint` SQL type.

is it intentional? 
What does this class represent actually?

charlesjmorgan replied:
My head was spinning dealing with the data types, I'll revisit the naming choices

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
```
That is mean we publish each page to kafka directly, as a set of rows?

- Is there any way to 'stage' produces messages and make them visible upon `finish` (or `ConnectorMetadata#finishInsert`)?
  - add a code comment explaining design decision
- should there be any buffering?

aalbu replied:
The producer does some buffering internally (configurable).  To make changes visible on `finishInsert()`, we'd have to start a transaction on `beginInsert()`, make sure that all producers share the same `transaction.id` (probably the Presto query id) and commit in `finishInsert()`.

I'm not a Kafka expert, so we felt we could use more time to think about this (feedback welcome) and some other issues.  That was intended for a follow-up PR (this one was supposed be sort of an MVP; it came out larger than I expected).

charlesjmorgan replied:
I'll look into this, also have to make it possible to send messages to different partitions

findepi replied:
> The producer does some buffering internally (configurable).

currently it's short-lived

kokosing replied:
Producer lives in worker while `finishInsert()` is called on coordinator.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/EncoderErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+
+    EncoderErrorCode(int code, ErrorType type)
+    {
+        errorCode = new ErrorCode(code + 0x0101_0000, name(), type);
```
the `0x0101_0000` number should be unique (it's not)

@electrum do you have a registry of these numbers or ?

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/EncoderFieldValueProvider.java:
```diff
@@ -0,0 +1,50 @@
+
+    ByteBuffer getVarbinary();
+
+    Slice getSlice();
+
+    Block getBlock();
```
these (and `getChar`) are unused

also, if the calling code could handle presto native types (slice, block), this class would be unnecessary

charlesjmorgan replied:
right, this class really doesn't do a whole lot I guess

findepi commented:
One non-obvious thing to me is about:
should we have set similar classes, interfaces, factories & providers for decoers & encoders 
OR should we enhance the decoders' interface to handle encoding as well.

i assume @charlesjmorgan @aalbu you considered that, so perhaps there is an answer ready.

charlesjmorgan commented:
@findepi 
The goal originally was to make an encoder that could be used by different connectors like the current record decoder. Any interfaces that seem unnecessary are there because of that.

Recently we changed the encoder to be Kafka specific so instead of returning a `byte []` it returns a `ProducerRecord` and then the serializers take care of... well serialization. @aalbu brought up what I think is a good point about keeping encoding and decoding separate and the single-responsibility principle. That is why while the `EncoderColumnHandle` might seem functionally equivalent to the `DecoderColumnHandle` they are two separate classes. If we were to combine the encoder and decoder in the future I think that there is a case to be made for just using one column handle.

I'm not sure what the best option is, keeping encoding and decoding separate or combining them. I'm also not sure if it would be better to have a one-size-fits-all encoder/decoder or make specific ones for each connector that might be optimized for each specific use case. Any thoughts you (or anyone else) have on this would be helpful.

aalbu commented:
> should we have set similar classes, interfaces, factories & providers for decoers & encoders

Perhaps.  I advocated for an iterative approach, where we implement the functionality for the connector at hand and then we can consider some abstractions that are more generic.  I feel that trying to generalize too early can lead to constraining decisions.

> OR should we enhance the decoders' interface to handle encoding as well.

So basically `RowDecoder` -> `RowCodec`?  That is a possibility, if we end up with symmetric abstractions.  I think that the `RowDecoder` could be generalized more (why do we assume that the source is a `byte[]`?).

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,62 @@
+        for (int i = 0; i < rowData.size(); ++i) {
+            EncoderFieldValueProvider fieldValueProvider = rowData.get(i);
+            row[i] = columnEncoders.get(fieldValueProvider.getColumn()).encodeField(fieldValueProvider);
+        }
```
nit: don't use `++` as an expression (saves reading time)

since you need `i`, an index `for` loop would be easier to read

findepi commented:
> I'm also not sure if it would be better to have a one-size-fits-all encoder/decoder or make specific ones for each connector that might be optimized for each specific use case

that's a very valid question. 
I am not sure whether decoder's reuse is the optimal choice, but we have it today anyway.

Since encoding seems pretty symmetric to decoding, it seems justified to bundle them in single class.
Doing so emphasizes this symmetry and may help understand the implementation.

I understand that, to realize this symmetry we would need kafka-independent types in the encoders interfaces.
byte[] or Slice seems like a safe choice. We would leave submitting this byte[] key/val to kafka topic in kafka-specific code.

> I think that the RowDecoder could be generalized more (why do we assume that the source is a byte[]?).

Because that's just convenient. You need to plug flexible interface when cost of copying data is large (ie data is large).
Current code is probably not optimized to avoid that, and I am not sure it should be.


@losipiuk please chime in, since you were working with this code more than I did.

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaProducerFactory.java:
```diff
@@ -0,0 +1,28 @@
+
+public interface KafkaProducerFactory
+{
+    default KafkaProducer<Object, Object> create(String keySerializer, String messageSerializer)
```
Can make return `KafkaProducer<byte[], byte[]>` and handle row encoding in separate layer like it is with `KafkaConsumerFactory`?

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,112 @@
+                messageEncoder);
+    }
+
+    private Map<String, String> getEncoderParameters(Optional<String> dataSchemaLocation)
```
returning `Map` here looks like an overkill

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaInsertTableHandle.java:
```diff
@@ -0,0 +1,153 @@
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaInsertTableHandle
```
If you would add columns to `KafkaTableHandle` then you could reuse it here.

Actually it is recommended that table handle should contain columns.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawColumnEncoder.java:
```diff
@@ -0,0 +1,278 @@
+            this.columnType = columnHandle.getType();
+
+            try {
+                fieldType = columnHandle.getDataFormat() == null ?
```
`Optional.ofNullable().map().  .orElse()?`?

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawColumnEncoder.java:
```diff
@@ -0,0 +1,278 @@
+                throw new IllegalArgumentException(format("invalid dataFormat '%s' for column '%s'", columnHandle.getDataFormat(), columnName));
+            }
+
+            String mapping = Optional.ofNullable(columnHandle.getMapping()).orElse("0");
```
this method is kinda long. One thing you may extract easily is `parseMapping`

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawColumnEncoder.java:
```diff
@@ -0,0 +1,278 @@
+            checkArgument(start >= 0, "start offset %s for column '%s' must be greater or equal 0", start, columnName);
+            end.ifPresent(endValue -> {
+                checkArgument(endValue >= 0, "end offset %s for column '%s' must be greater or equal 0", endValue, columnName);
+                checkArgument(endValue >= start, "end offset %s for column '%s' must greater or equal start offset", endValue, columnName);
```
I wonder if equal makes sense

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawColumnEncoder.java:
```diff
@@ -0,0 +1,278 @@
+            checkArgument(isSupportedType(columnType), "Unsupported column type '%s' for column '%s'", columnType.getDisplayName(), columnName);
+
+            if (!isVarcharType(columnType)) {
+                checkArgument(!end.isPresent() || end.getAsInt() - start == fieldType.getSize(),
```
can you validate agains `length` here?

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroColumnEncoder.java:
```diff
@@ -0,0 +1,106 @@
+            BIGINT,
+            DOUBLE,
+            REAL,
+            VARBINARY);
```
varbinary is parametrized with lengh

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroColumnEncoder.java:
```diff
@@ -0,0 +1,106 @@
+        else if (columnType == REAL) {
+            record.put(columnName, fieldValueProvider.getReal());
+        }
+        else if (columnType == VARBINARY) {
```
isVarbinaryType

aalbu commented:
> I understand that, to realize this symmetry we would need kafka-independent types in the encoders interfaces.
> byte[] or Slice seems like a safe choice. We would leave submitting this byte[] key/val to kafka topic in kafka-specific code.

But what's the value of that?  There isn't much code reuse.  It's not an intuitive way of working with target systems.  We might make Kafka work with it, but do we know that it's a good fit for other systems?  Redis is using the `RecordDecoder` - can we make Redis writes work by providing it byte arrays?  Would it work when you persist a list?  I don't think we have given that much thought, so I feel there is no supporting evidence that extending the interface with encoding methods producing a 'byte[]` is appropriate.

findepi commented:
> But what's the value of that? There isn't much code reuse.

This isn't about code reuse, but more about physically bundling things that are coupled.
For example, i would want `Integer#parseInt` and `Integer#toString` conversions to be implemented 
in one class, as one is reverse of the other (and they are not so complex to preclude being one class).

> Redis is using the RecordDecoder - can we make Redis writes work by providing it byte arrays?

I don't know. _IF_ the decoding was accepting `byte[]` _THEN_ it would be reasonable for the encoding to return `byte[]`.
However, I just realized this is worse than that because decoding takes `byte[] data, Map<String, String> dataMap`
(and the `dataMap` is specific to redis).
That suggests redis is not a good use-case for byte[]-based row decoding/encoding at all.

losipiuk commented:
> This isn't about code reuse, but more about physically bundling things that are coupled

I agree that bundling encoders and decoders in single class would make reading and understanding implementation easier. Though it will push us more towards reusing same logic in very much different connectors. I think the fact that Kafka and Redis are sharing decoder implementation does more harm than good (see example @findepi gave above). 

I think (if we find resources for that)  we should untangle the situation and do either:

a) rework presto-decoder module to make it more connector agnostic and keep it reused where we currently use it. Then we can combine `PrestoDecoder` with `PrestoEncoder` in the shared module.

b) stop using presto-decoder in connectors it does not play well with (Redis?). This can potentially result in dropping presto-decoder at all if it looks like having separate implementation for each connector is better. Then we can combine decoder and encoder logic in Kafka connector module.

Short term probably it does not matter much. At least for this PR I would keep them separate so we can focus on merging it sooner. Then as next step (we should not postpone that) we can work on refactor and move it towards either a) or b). WDYT?

findepi commented:
FWIW, the `presto-record-decoder` module is currently used in:

- Kafka
- Redis 
- Kinesis

> At least for this PR I would keep them separate so we can focus on merging it sooner. 

@losipiuk not sure what you mean here. How does separate classes vs extending existing classes matter?

FYI i do not have very strong opinion either way, but i expect it may be some work to change between approaches.

losipiuk commented:
> @losipiuk not sure what you mean here. How does separate classes vs extending existing classes matter?

It matters in terms how many rounds of review we need. I would prefer to not bloat this PR with extra refactorings. Smaller PRs are much easier to review and work with. And sooner we merge this beast and continue with smaller gradual changes the better IMO.

charlesjmorgan commented:
Thank you @aalbu @findepi @losipiuk and @kokosing for all your feedback, it has been very helpful! I hope I got everything (I triple checked so should be good). I intentionally left out changes that I didn't think were in the scope of this PR, I will revisit those in the future. I am going to split this pr into 5 parts. The first will be basic functionality for inserts and the four after that will each be for a specific encoder format. lmk if you think I should split it up differently
